File: backend/internal/scheduler/sjf.go
package scheduler

import (
	"github.com/pdro1812/trabalho-os-scheduler/backend/internal/models"
)

func RunSJFNonPreemptive(req models.SimulationRequest) models.SimulationResponse {
	processes := make([]models.Process, len(req.Processes))
	copy(processes, req.Processes)

	var executionOrder []models.ExecutionBlock
	var metrics []models.ProcessMetrics

	currentTime := 0
	completed := 0
	n := len(processes)
	isCompleted := make([]bool, n)

	totalWaitTime := 0
	totalTurnaroundTime := 0

	for completed < n {
		idx := -1
		minBurst := 999999999 // Valor alto para servir de infinito

		// Procura o processo que já chegou e tem o menor tempo de CPU
		for i := 0; i < n; i++ {
			if !isCompleted[i] && processes[i].ArrivalTime <= currentTime {
				if processes[i].BurstTime < minBurst {
					minBurst = processes[i].BurstTime
					idx = i
				}
				// Critério de desempate caso dois tenham o mesmo burst time: quem chegou primeiro ganha
				if processes[i].BurstTime == minBurst && idx != -1 {
					if processes[i].ArrivalTime < processes[idx].ArrivalTime {
						idx = i
					}
				}
			}
		}

		if idx != -1 {
			// Encontramos o menor processo, vamos rodá-lo
			p := processes[idx]
			start := currentTime
			currentTime += p.BurstTime
			end := currentTime

			executionOrder = append(executionOrder, models.ExecutionBlock{
				PID:   p.PID,
				Start: start,
				End:   end,
			})

			turnaround := end - p.ArrivalTime
			wait := turnaround - p.BurstTime

			metrics = append(metrics, models.ProcessMetrics{
				PID:           p.PID,
				EffectiveTime: turnaround,
				WaitingTime:   wait,
			})

			totalTurnaroundTime += turnaround
			totalWaitTime += wait
			isCompleted[idx] = true
			completed++

			// Aplica TTC
			if completed < n && req.TTC > 0 {
				executionOrder = append(executionOrder, models.ExecutionBlock{
					PID:   "TTC",
					Start: currentTime,
					End:   currentTime + req.TTC,
				})
				currentTime += req.TTC
			}

		} else {
			// CPU Ociosa: Ninguém chegou ainda. Avança o tempo para a chegada do próximo.
			nextTime := 999999999
			for i := 0; i < n; i++ {
				if !isCompleted[i] && processes[i].ArrivalTime < nextTime {
					nextTime = processes[i].ArrivalTime
				}
			}
			currentTime = nextTime
		}
	}

	return models.SimulationResponse{
		ExecutionOrder:    executionOrder,
		Metrics:           metrics,
		AvgWaitTime:       float64(totalWaitTime) / float64(n),
		AvgTurnaroundTime: float64(totalTurnaroundTime) / float64(n),
	}
}

File: backend/internal/scheduler/sjf.js
function runSJFNonPreemptive(requisicao) {
    const processos = [];
    for (let i = 0; i < requisicao.Processes.length; i++) {
        processos.push(requisicao.Processes[i]);
    }

    const n = processos.length;
    const ordemExecucao = [];
    const metricas = [];
    let tempoAtual = 0;
    let processosConcluidos = 0;

    // Array para marcar quais processos já terminaram
    const concluido = new Array(n).fill(false);

    let somaEspera = 0;
    let somaTurnaround = 0;

    // Ajuda a saber se a CPU estava parada (ociosa)
    let veioDeOciosidade = true;

    while (processosConcluidos < n) {
        let indiceEscolhido = -1;
        let menorBurst = Infinity;

        // --- PASSO 1: Selecionar o próximo processo ---
        // Procuramos o processo que já chegou e tem o menor tempo de execução (BurstTime)
        for (let i = 0; i < n; i++) {
            // Só olhamos processos que ainda não terminaram E que já chegaram no tempo atual
            if (concluido[i] === false && processos[i].ArrivalTime <= tempoAtual) {

                // Critério: menor BurstTime
                if (processos[i].BurstTime < menorBurst) {
                    menorBurst = processos[i].BurstTime;
                    indiceEscolhido = i;
                } 
                // Se houver empate no BurstTime, escolhemos o que chegou primeiro
                else if (processos[i].BurstTime === menorBurst) {
                    if (processos[i].ArrivalTime < processos[indiceEscolhido].ArrivalTime) {
                        indiceEscolhido = i;
                    }
                }
            }
        }

        // --- PASSO 2: Executar o processo escolhido ---
        if (indiceEscolhido !== -1) {
            const p = processos[indiceEscolhido];

            // Se não veio de ociosidade, cobramos o tempo de troca de contexto (TTC)
            if (veioDeOciosidade === false && requisicao.TTC > 0) {
                ordemExecucao.push({
                    PID: "TTC",
                    Start: tempoAtual,
                    End: tempoAtual + requisicao.TTC
                });
                tempoAtual = tempoAtual + requisicao.TTC;
            }

            // Agora o processo roda até o fim (não-preemptivo)
            veioDeOciosidade = false;
            const inicio = tempoAtual;
            tempoAtual = tempoAtual + p.BurstTime;
            const fim = tempoAtual;

            ordemExecucao.push({
                PID: p.PID,
                Start: inicio,
                End: fim
            });

            // Calculamos as métricas
            const turnaround = fim - p.ArrivalTime;
            const espera = turnaround - p.BurstTime;

            metricas.push({
                PID: p.PID,
                EffectiveTime: turnaround,
                WaitingTime: espera
            });

            somaTurnaround += turnaround;
            somaEspera += espera;

            // Marcamos como concluído para não escolher ele de novo
            concluido[indiceEscolhido] = true;
            processosConcluidos++;

        } else {
            // --- PASSO 3: Lidar com a ociosidade ---
            // Se nenhum processo chegou ainda, avançamos o tempo para a chegada do próximo processo
            let proximaChegada = Infinity;
            for (let i = 0; i < n; i++) {
                if (concluido[i] === false && processos[i].ArrivalTime < proximaChegada) {
                    proximaChegada = processos[i].ArrivalTime;
                }
            }

            tempoAtual = proximaChegada;
            veioDeOciosidade = true;
        }
    }

    return {
        ExecutionOrder: ordemExecucao,
        Metrics: metricas,
        AvgWaitTime: somaEspera / n,
        AvgTurnaroundTime: somaTurnaround / n
    };
}

module.exports = { runSJFNonPreemptive };